.cirrus.yml

```diff
@@ -0,0 +1,121 @@
+common_steps_template: &COMMON_STEPS_TEMPLATE
+  install_prerequisites_script: |
+    set -uexo pipefail
+    if [ "$OS_NAME" == "linux" ]; then
+      packages="git-core make g++ gdb curl libcurl3 tzdata zip unzip xz-utils"
+      if [ "$MODEL" == "32" ]; then
+        dpkg --add-architecture i386
+        packages="$packages g++-multilib libcurl3-gnutls:i386"
+      fi
+      if [ "${DMD:0:4}" == "gdmd" ]; then
+        packages="$packages sudo software-properties-common wget"
+      fi
+      apt-get -q update
+      apt-get install -yq $packages
+    elif [ "$OS_NAME" == "darwin" ]; then
+      # required for install.sh
+      brew install gnupg
+    elif [ "$OS_NAME" == "freebsd" ]; then
+      pkg install -y git gmake bash
+      rm /usr/bin/make
+      ln -s /usr/local/bin/gmake /usr/bin/make
+    fi
+    # create a `dmd` symlink to the repo dir, necessary for druntime/Phobos
+    ln -s $CIRRUS_WORKING_DIR ../dmd
+  install_host_compiler_script: |
+    source ci.sh
+    # kludge
+    if [ "${DMD:0:4}" == "gdmd" ]; then export DMD="gdmd"; fi
+    install_d "$DMD"
+  setup_repos_script: |
+    export BRANCH=${CIRRUS_BASE_BRANCH:-$CIRRUS_BRANCH}
+    source ci.sh
+    setup_repos
+  build_script: |
+    source ci.sh
+    build
+  test_dmd_script: |
+    if [ -z "$CIRRUS_PR" ]; then export FULL_BUILD="true"; fi
+    source ci.sh
+    test_dmd
+  test_druntime_script: |
+    set -uexo pipefail
+    make -j$N -C ../druntime -f posix.mak MODEL=$MODEL unittest
+  test_phobos_script: |
+    set -uexo pipefail
+    make -j$N -C ../phobos -f posix.mak MODEL=$MODEL unittest
+
```

Comment:
I don't know much yaml but I read it has the ability to run tasks concurrently - do we test dmd, druntime, and phobos in parallel?

Reply:
No this is sequential (like on almost all of our other CIs). We only do parallel testing on Buildkite (at the moment). Typically >95% of the time is spent DMD's testsuite, so the coolest would be to split the testsuite one off (though that's definitely for future improvements).

Reply:
For Cirrus, a 'task' is a job, and they all run simultaneously if the (per-user, not per repo!) limits aren't exceeded. With the current setup, all jobs/tasks are schedulable at once for a user; for FreeBSD, another job with 4 CPU cores would still be available.

Each of these steps is sequential inside a task, but each step tries to use full parallelization. That works quite well for the DMD tests and compiling the druntime/Phobos unittest modules; installing prerequisites etc. doesn't benefit from more CPU cores. Parallelizing these steps isn't possible with Cirrus AFAIK, and could lead to out-of-memory errors due to excessive threads.

Cirrus usually shows nice graphs wrt. CPU and memory usage, see e.g. https://cirrus-ci.com/task/5287685575147520 for LDC. Not sure why it doesn't show up here, maybe it needs to be merged first.